tools/pythonpkg/src/pandas_analyzer.cpp

```diff
@@ -0,0 +1,379 @@
+#include "duckdb_python/pyrelation.hpp"
+#include "duckdb_python/pyconnection.hpp"
+#include "duckdb_python/pyresult.hpp"
+#include "duckdb_python/pandas_analyzer.hpp"
+#include "duckdb_python/python_conversion.hpp"
+
+namespace duckdb {
+
+static bool UpgradeType(LogicalType &left, const LogicalType &right);
+
+static bool SameTypeRealm(LogicalTypeId a, LogicalTypeId b) {
+	if (a == b) {
+		return true;
+	}
+	if (a > b) {
+		return SameTypeRealm(b, a);
+	}
+	D_ASSERT(a < b);
+
+	// anything ANY and under can transform to anything
+	if (a <= LogicalTypeId::ANY) {
+		return true;
+	}
+	// Both a and b are not nested
+	if (b < LogicalTypeId::STRUCT) {
```

Comment:
`ENUM` is actually not nested as well, and is after `LIST` in the logical type enum. Perhaps we could do a comparison to the various nested types in a separate function (e.g. `TypeIsNested`)?

Reply:
Yea good idea, JSON is also a nested type, no?

So the list of nested types are:
- STRUCT
- LIST
- MAP
- JSON
- AGGREGATE_STATE (?)
- TABLE

Is that list correct, am I missing any, or are some actually not nested? (I'm not sure about AGGREGATE_STATE)

Reply:
It is only STRUCT, LIST and MAP. TABLE is unused, and aggregate_state is a blob.